### ISSUE TITLE
Early Stopping Min Epochs

### ISSUE THREAD

**Issue by thomasj02** [PyTorchLightning/pytorch-lightning#6446]
## 🚀 Feature
The EarlyStopping callback should allow the user to specify a minimum number of epochs to run before early stopping is triggered

### Motivation

In many modern training loops, the learning rate is varied in some kind of cycle. For example, in the Transformer paper, they warm up the learning rate by increasing it gradually until a certain number of training steps, then they cool it by decreasing it as 1/sqrt(steps). 

In cases like these, the user may want to run a certain minimum number of epochs, for example to ensure that the learning rate has achieved a given value before stopping.

### Pitch

An additional parameter, `min_epochs` should be added to EarlyStopping (default=0 for backwards compatibility). The stopping will then trigger if both `patience` epochs have passed since the last improvement and at least `min_epochs` epochs have passed in total.

**Comment by carmocca** [comment]
I like it, feel free to open a PR!

**Comment by ananthsub** [comment]
As a general feature this sounds good. 
@carmocca if `min_epochs` is configured on the Trainer, does EarlyStopping override that? What's the expected priority between Trainer settings and callback settings like this?

**Comment by carmocca** [comment]
Turns out that `Trainer.min_epochs` is used to override the EarlyStopping behaviour, just as OP asked.

https://github.com/PyTorchLightning/pytorch-lightning/blob/f4cc7451a94010a572480c43ad5f0af7ad52cd21/pytorch_lightning/trainer/trainer.py#L594-L606

where `self.should_stop` is written by `EarlyStopping`

But it makes more sense to me that `EarlyStopping.min_epochs` would be used for that and instead `Trainer.min_epochs` would define a minimum number of epochs to run before stopping, regardless of EarlyStopping.

**Comment by hasan-rakibul** [comment]
> Turns out that `Trainer.min_epochs` is used to override the EarlyStopping behaviour, just as OP asked.
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/f4cc7451a94010a572480c43ad5f0af7ad52cd21/pytorch_lightning/trainer/trainer.py#L594-L606
> 
> where `self.should_stop` is written by `EarlyStopping`
> 
> But it makes more sense to me that `EarlyStopping.min_epochs` would be used for that and instead `Trainer.min_epochs` would define a minimum number of epochs to run before stopping, regardless of EarlyStopping.

I think it does not satisfy what OP asked. What I found, the early stopping is just _delayed_ until `min_epochs` satisfies. What OP requested (and I request as well) is to trigger early stopping after a certain number of epochs.

Here are some logs from my experiment; it ends training right on Epoch 4 (although it shows improvement on Epoch 4), presumably because the early stopping condition was met at Epoch 2.

```
Epoch 2: 100%|██████████| 259/259 [00:26<00:00,  9.84it/s, v_num=1.83e+7]Monitored metric val_pcc did not improve in the last 2 records. Best score: 0.266. Signaling Trainer to stop.
   
Epoch 3:   0%|          | 0/259 [00:00<?, ?it/s, v_num=1.83e+7]Trainer was signaled to stop but the required `min_epochs=5` or `min_steps=None` has not been met. Training will continue...

Epoch 4: 100%|██████████| 259/259 [00:27<00:00,  9.47it/s, v_num=1.83e+7]Metric val_pcc improved by 0.244 >= min_delta = 0.001. New best score: 0.693
```
